src/main/java/org/folio/circulation/services/LostItemFeeChargingService.java

```diff
@@ -54,35 +71,126 @@ public LostItemFeeChargingService(Clients clients) {
     this.locationRepository = LocationRepository.using(clients);
     this.eventPublisher = new EventPublisher(clients.pubSubPublishingService());
     this.feeFineScheduledNoticeService = FeeFineScheduledNoticeService.using(clients);
+    this.refundService = new LostItemFeeRefundService(clients);
+    this.accountRepository = new AccountRepository(clients);
+    this.feeFineService = new FeeFineService(clients);
+    this.userRepository = new UserRepository(clients);
   }
 
   public CompletableFuture<Result<Loan>> chargeLostItemFees(
     Loan loan, DeclareItemLostRequest request, String staffUserId) {
+    this.userId = staffUserId;
+    this.servicePointId = request.getServicePointId();
 
     final ReferenceDataContext referenceDataContext = new ReferenceDataContext(
       loan, request, staffUserId);
-
     return lostItemPolicyRepository.getLostItemPolicyById(loan.getLostItemPolicyId())
       .thenApply(result -> result.map(referenceDataContext::withLostItemPolicy))
       .thenCompose(refDataResult -> refDataResult.after(referenceData -> {
         if (shouldCloseLoan(referenceData.lostItemPolicy)) {
           log.debug("Loan [{}] can be closed because no fee will be charged", loan.getId());
-          return closeLoanAndUpdateInStorage(loan)
+          return closeLoanAsLostAndPaidAndUpdateInStorage(loan)
             .thenCompose(r -> r.after(eventPublisher::publishClosedLoanEvent))
             .thenApply(r -> r.map(v -> loan));
         }
 
-        return fetchFeeFineOwner(referenceData)
-          .thenApply(this::refuseWhenFeeFineOwnerIsNotFound)
-          .thenComposeAsync(this::fetchFeeFineTypes)
-          .thenApply(this::buildAccountsAndActions)
-          .thenCompose(r -> r.after(feeFineFacade::createAccounts))
-          .thenApply(r -> r.map(notUsed -> loan));
+        log.info("Checking for existing lost item fees for loan [{}]", loan.getId());
+        return accountRepository.findAccountsForLoan(loan)
```

Comment:
I'd personally don't change this class (because it couples fee charging and refund logic now), but place the logic in the `DeclareLostResource` (assuming it is declared lost specific logic). 

I believe we even don't need all these changes (but I must missing something), we can do something like:
* Before line # 71 in `DeclareLostResource` class just add a call to  `org.folio.circulation.services.LostItemFeeRefundService#refundLostItemFees(org.folio.circulation.services.LostItemFeeRefundContext)` (we have to make it public) - the method should cancel/refund all *lost* fees for the item, if any.

The only possible issue here is whether we should cancel the fees even after refund period? If yes, than we have to update the refund code a bit so it is cover this case (maybe add a predicate to the `LostItemFeeRefundContext` - which is always true for your case, but `lostItemPolicy.shouldRefundFees(...)` by default).

Reply:
Thanks @bohdan-suprun for the advice on how to structure this process

Reply:
>I'd personally don't change this class (because it couples fee charging and refund logic now), but place the logic in >the DeclareLostResource (assuming it is declared lost specific logic).

This is the first place I attempted to put it.  The problem I ran into is that the account management functions are defined in a separate package and inaccessible from there.  This is why I ended up putting it in the fee charging step.

>org.folio.circulation.services.LostItemFeeRefundService#refundLostItemFees(org.folio.circulation.services.LostItem>FeeRefundContext) (we have to make it public) - the method should cancel/refund all lost fees for the item, if any.

That is exactly how I had the code written originally.  During testing, what I found is that this service does not, in fact, cancel all fees. What it actually seems to do is cancel partially/fully paid or transferred accounts.  In fact, if it operates on a loan that *only* has unpaid/untransferred accounts, it fails the returned result with an error message of "could not find accounts to refund."  Which is why I built separate cancellation code, and why I have a separate account check at the end of the process, instead of checking the success or failure of the call to the refund service.

Reply:
> "could not find accounts to refund."

Do you know which line causes this error? This should be regression/some missing settings. There must be an existing test case that demonstrates this behavior. (e.g. `api.loans.scenarios.RefundDeclaredLostFeesTestBase#shouldCancelBothItemAndProcessingFeesWhenNeitherHaveBeenPaid`)

Reply:
I don't, because I am not familiar with this code and just assumed that was how the service is supposed to work.  I'm happy to do some more digging and see if I can trace the problem. It would really, really simplify my code if this service actually did close/cancel everything and I could depend on a single meaning for a failure.

Reply:
Okay, I looked at this today-I think the relevant code for this behavior is in the FeeFineFacade class, in the RefundAccountIfneeded method.  If you trace the code back from the LostItemFeeRefundService#refundLostItemFees class, this seems to be what it's using to actually close the accounts.  On line 92 of that file you'll find this code:

    if (!account.hasPaidOrTransferredAmount()) {
      log.info("Account {} is not processed yet, no refunds have to be issued", account.getId());
      return ofAsync(() -> null);
    }
So that method is explicitly and intentionally checking for accounts that have been partially paid or transferred, and kicking out accounts that don't fit that criteria-which is why the refund service is not closing all accounts.